torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+import distutils.util
+import os
+import torch
+import torch.distributed as dist
+import torch_xla
+import torch_xla.core.xla_model as xm
+import logging
+from torch._C._distributed_c10d import (
+    ProcessGroup,
+    Work,
+)
+
+
+def _create_xla_process_group(prefix_store, rank, size, timeout):
+  return ProcessGroupXla(prefix_store, rank, size, timeout)
+
+
+def _register_xla_backend():
+  dist.Backend.register_backend('xla', _create_xla_process_group)
+
+
+_register_xla_backend()
+
+
+class ProcessGroupXla(ProcessGroup):
+  '''ProcessGroup for XLA devices. See ProcessGroup for doc.
+
+    https://github.com/pytorch/pytorch/blob/074c77601114c66df346c5465667ce8cfaa6d22c/torch/csrc/distributed/c10d/ProcessGroup.hpp#L79
+    https://github.com/pytorch/pytorch/blob/074c77601114c66df346c5465667ce8cfaa6d22c/torch/csrc/distributed/c10d/init.cpp#L937
+    '''
+
+  def __init__(self, prefix_store, rank, size, timeout):
+    super().__init__(rank, size)
+    self.prefix_store = prefix_store  # reserved for future use.
+    self.timeout = timeout
+    self._mesh = []
+
+  def _get_reduce_type(self, reduce_op):
+    if reduce_op == dist.ReduceOp.SUM:
+      return xm.REDUCE_SUM
+    elif reduce_op == dist.ReduceOp.PRODUCT:
+      return xm.REDUCE_MUL
+    elif reduce_op == dist.ReduceOp.BAND:
+      return xm.REDUCE_AND
+    elif reduce_op == dist.ReduceOp.BOR:
+      return xm.REDUCE_OR
+    elif reduce_op == dist.ReduceOp.MIN:
+      return xm.REDUCE_MIN
+    elif reduce_op == dist.ReduceOp.MAX:
+      return xm.REDUCE_MAX
+    elif reduce_op == dist.ReduceOp.BXOR:
+      raise NotImplementedError(f'reduce op {reduce_op}')
+    else:
+      raise ValueError(f'Invalid reduce op {reduce_op}')
+
+  def allreduce(self, tensors, all_reduce_options):
+    reduce_type = self._get_reduce_type(all_reduce_options.reduceOp)
+
+    # TODO(junminh): implement all_reduce_options.timeout.
+    xm.all_reduce(reduce_type, tensors, groups=self._mesh)
```

Comment:
Curious, will this run immediately and blocking or does it also run lazily until `mark_step` is called??

Reply:
`xm.all_reduce` only builds the graph and only be executed when a `mark_step` is called. I expect this new distributed API to share the same behavior.